asllib/aslspec/spec.ml

```diff
@@ -251,61 +344,45 @@ type t = {
 
 let ast spec = spec.ast
 
-(** [defining_node_for_id self id] returns the defining node for the element
-    with identifier [id] that is given in the specification. If no such element
-    exists, a [SpecError] is raised. *)
-let defining_node_for_id self id =
-  try StringMap.find id self.id_to_defining_node
-  with Not_found ->
-    let msg = Format.sprintf "Encountered undefined element: %s" id in
-    raise (SpecError msg)
-
 module Check = struct
-  let list_find_duplicate strs =
-    let module StringSet = Set.Make (String) in
-    let rec find_with_set set = function
-      | [] -> None
-      | str :: tail ->
-          if StringSet.mem str set then Some str
-          else find_with_set (StringSet.add str set) tail
-    in
-    find_with_set StringSet.empty strs
-
+  (** [check_layout term layout] checks that the given [layout] is structurally
+      consistent with the given [term]. If not, raises a [SpecError] describing
+      the issue. *)
   let rec check_layout term layout =
-    let msg =
-      Format.asprintf
-        "layout %a is inconsistent with %a. Here's a consistent layout: %a"
-        PP.pp_math_shape layout PP.pp_type_term term PP.pp_math_shape
-        (Layout.default_for_type_term term)
-    in
+    let consistent_layout = Layout.default_for_type_term term in
     let open Layout in
     match (term, layout) with
     | Label _, Unspecified -> ()
-    | Label _, _ -> raise (SpecError msg)
+    | Label _, _ -> Error.bad_layout term layout consistent_layout
     | TypeOperator { term = _, t }, _ -> check_layout t layout
     | LabelledTuple { components }, Horizontal cells
     | LabelledTuple { components }, Vertical cells ->
         if List.length components <> List.length cells then
```

Comment:
nit: consider using `List.compare_lengths`

```suggestion
        if List.compare_lengths components cells <> 0 then
```

There are at least 4 occurrences of this pattern